dipy/align/crosscorr.pyx

```diff
@@ -27,17 +28,115 @@ cdef enum:
     SIJ = 4
     CNT = 5
 
+
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
-def precompute_cc_factors_3d(floating[:, :, :] static, floating[:, :, :] moving,
-                             cnp.npy_intp radius):
+cdef inline int _wrap(int x, int m)nogil:
+    r""" Auxiliary function to `wrap` an array around its low-end side
+    Negative indices are mapped to last coordinates so that no extra memory
+    is required to account for local rectangular windows that exceed the
+    array's low-end boundary.
+
+    Parameters
+    ----------
+    x : int
+        the array position to be wrapped
+    m : int
+        array length
+    """
+    if x<0:
+        return x + m
+    return x
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef inline void _update_factors(double[:,:,:,:] factors,
+                                 floating[:,:,:] moving,
+                                 floating[:,:,:] static,
+                                 int ss, int rr, int cc,
+                                 int s, int r, int c, int operation)nogil:
+    r"""Updates the precomputed CC factors of a rectangular window
+
+    Updates the precomputed CC factors of the rectangular window centered
+    at (`ss`, `rr`, `cc`) by adding the factors corresponding to voxel
+    (`s`, `r`, `c`) of input images `moving` and `static`.
+
+    Parameters
+    ----------
+    factors : array, shape (S, R, C, 5)
+        array containing the current precomputed factors to be updated
+    moving : array, shape (S, R, C)
+        the moving volume (notice that both images must already be in a common
+        reference domain, i.e. the same S, R, C)
+    static : array, shape (S, R, C)
+        the static volume, which also defines the reference registration domain
+    ss : int
+        first coordinate of the rectangular window to be updated
+    rr : int
+        second coordinate of the rectangular window to be updated
+    cc : int
+        third coordinate of the rectangular window to be updated
+    s: int
+        first coordinate of the voxel the local window should be updated with
+    r: int
+        second coordinate of the voxel the local window should be updated with
+    c: int
+        third coordinate of the voxel the local window should be updated with
+    operation : int, either -1, 0 or 1
+        indicates whether the factors of voxel (`s`, `r`, `c`) should be
+        added to (`operation`=1), subtracted from (`operation`=-1), or set as
+        (`operation`=0) the current factors for the rectangular window centered
+        at (`ss`, `rr`, `cc`).
+
+    """
+    cdef:
+        double sval
+        double mval
+    if s>=moving.shape[0] or r>=moving.shape[1] or c>=moving.shape[2]:
+        if operation == 0:
+            factors[ss, rr, cc, SI] = 0
```

Comment:
As far as I know, cython enum elements are always integer, like in C. I did a quick search to see if there exists a more general enum implementation using different element types (who knows! people are very creative! =) ), but I couldn't find any.